### ISSUE TITLE
[Improvement] Better version tracking between git and nuget

### ISSUE THREAD

**Issue by knocte** [xamarin/Xamarin.Forms#7841]
## Summary
Currently, versions used in nuget packages don't provide any clue whatsoever about what exact version in git they map to. This makes it damn hard for contributors to work on bugfixes that fix regressions, because the easiest way to find the culprit of a regression is to bisect between 2 versions (1 bad and 1 good).

Example: the latest releases in nuget are 4.3.0.851321-pre3 and 4.2.0.848062, while in git we have the git tag `release-4.2.0-sr3` (and no git tags for 4.3.x prereleases).

## Proposed Changes
- For prereleases, include the git hash in the nuget version (an example of this can be found on this nuget package that I own: https://www.nuget.org/packages/JsonRpcSharp; as you can see, not only the git hash is included, but also the date, which is helpful but most importantly keeps the newest-comes-first sorting in nuget homepage).
- For normal releases, including the git hash in the version without marking the release as pre-release would be, AFAIU, hard if not impossible, so at least I propose in this case that git tag names map exactly to the version released in nuget.

## Intended Use Case
Working with this versioning model will allow contributors be more productive tracking&fixing regressions and stop thinking of nuget packages/versions as a blackbox asset.

**Comment by adrianknight89** [comment]
I struggled with this in the past. This would be a great change. I'd like to see the git hash in the nightly builds as well. If adding the hashes to Nuget names is not possible, then I'd like to see a wiki with the 1:1 mapping & proper links. In addition, I'd like to know which branch was used to build a Nuget package. Thus, we need a table of _versions, branches, and hashes_.

> For normal releases, including the git hash in the version without marking the release as pre-release would be, AFAIU, hard if not impossible

Can you explain why you think it'd be hard?

**Comment by knocte** [comment]
> If adding the hashes to Nuget names is not possible

It is possible, look at the link I provided, it's done there in that nuget package.

> Can you explain why you think it'd be hard?

AFAIU, letters in version numbers make the version be a pre-release. Not sure if this is a rule in the SemVer spec.

**Comment by adrianknight89** [comment]
I saw the link before. What I meant was whether the team wanted to go that route or not depending on other factors.

**Comment by jfversluis** [comment]
Hi @knocte thanks for the suggestion!

I think it might be much easier than you think. Did you check out the [Releases tab](https://github.com/xamarin/Xamarin.Forms/releases) on our repo? :) Does that already solve some of the pain you're experiencing?

**Comment by knocte** [comment]
> Does that already solve some of the pain you're experiencing?

Oh wow, didn't know that this tab contained very useful data (such as the git hash of each release on the left side). But it could be improved because you don't see the release number that gets published in nuget unless you click on the "Assets" part to see each file.

Also it didn't help that it seems github doesn't show all the tags of the repository in the dropdown of the first landing page of the project (all the tags starting with the word "beta" are not there, why), while the page https://github.com/xamarin/Xamarin.Forms/tags contains all.

**Comment by jfversluis** [comment]
> Also it didn't help that it seems github doesn't show all the tags of the repository in the dropdown of the first landing page of the project (all the tags starting with the word "beta" are not there, why?)

I'm not entirely sure what you mean by this, sorry.

**Comment by knocte** [comment]
> I'm not entirely sure what you mean by this, sorry.

<img width="381" alt="Screenshot 2019-10-13 at 6 52 39 PM" src="https://user-images.githubusercontent.com/331303/66714579-ca14c600-edea-11e9-8f6c-27dd213b4f8c.png">

**Comment by jfversluis** [comment]
Ah, right. I'm guessing that GitHub does not show the tags in there that are marked as pre-release. Unfortunately, that is not something we can influence.

**Comment by knocte** [comment]
I'm guessing we could write some bits about all this in the CONTRIBUTING.md file?

**Comment by jfversluis** [comment]
@knocte about the releases tab you mean? Definitely!

**Comment by knocte** [comment]
Yeah, and the nuget version numbers in the assets sections, and the pre-release git tags, etc
